Add WindowsTracing project

pierricgimmig opened the pull request:
WindowsTracing uses the recently introduced krabsetw library to access 
"Event Tracing for Windows" (ETW) events. Currently, only scheduling 
events and thread information events are enabled.

florian-kuebler commented on src/WindowsTracing/ContextSwitchManagerTest.cpp:
```diff
@@ -0,0 +1,109 @@
+  orbit_windows_tracing::MockTracerListener mock_listener;
+  ContextSwitchManager manager(&mock_listener);
+
+  EXPECT_CALL(mock_listener, OnSchedulingSlice).Times(testing::Exactly(1));
```
Nit: you can simply state `Times(1)`.
Also, I would also add do a `SaveArg`, to check that the scheduling slice contains the correct information.

Both applies for here and below.

florian-kuebler replied:
Edit (since I read the code below and the Mock class):
Ok, I see using the DelegateToFake. I think that is fine. `SaveArg` would be probably my choice here, but that's personal preference.

However, I'd do the actual check for the values of the scheduling slice in this test anyways. Or just remove this test case, as it does not gain any more information than the one next.

pierricgimmig replied:
The pattern comes from here http://google.github.io/googletest/gmock_cook_book.html#DelegatingToFake. Will see if SaveArg is simpler.

pierricgimmig replied:
Removed `FakeTracerListener` in favor of SaveArg.

florian-kuebler commented on src/WindowsTracing/ContextSwitchManagerTest.cpp:
```diff
@@ -0,0 +1,109 @@
+  EXPECT_EQ(fake_listener.scheduling_slices_.size(), 1);
+  EXPECT_EQ(fake_listener.scheduling_slices_[0].pid(), 3);
+}
+
```
I'd probably add another test, that checks for multiple scheduling slices.

pierricgimmig replied:
Done

florian-kuebler commented on src/WindowsTracing/include/WindowsTracing/TracerListener.h:
```diff
@@ -0,0 +1,20 @@
+class TracerListener {
+ public:
+  virtual ~TracerListener() = default;
+  virtual void OnSchedulingSlice(orbit_grpc_protos::SchedulingSlice scheduling_slice) = 0;
```
Long term, we should probably have one listener for both, windows and linux tracing.

pierricgimmig replied:
My initial thought was to go for a single listener since to reuse code from the Linux OrbitService, but in the end I agree that it's cleaner not to.

florian-kuebler commented on src/WindowsTracing/include/WindowsTracing/Tracer.h:
```diff
@@ -0,0 +1,31 @@
+
+namespace orbit_windows_tracing {
+
+class Tracer {
```
Please add a brief class comment.

pierricgimmig replied:
Done, thanks.

florian-kuebler commented on src/WindowsTracing/Tracer.cpp:
```diff
@@ -0,0 +1,28 @@
+
+void Tracer::Start() {
+  CHECK(tracer_impl_ == nullptr);
+  tracer_impl_ = std::make_unique<KrabsTracer>(capture_options_, listener_);
```
I am a bit confused about this implementation. Shouldn't Tracer be simply an interface, and we have KrabsTracer as subclass, implementing Start/Stop, and maybe provide a Factory for the concrete tracer?

dpallotti replied:
Yeah, `tracer_impl_` is a field holding an instance of a subclass, something seems wrong. Either `KrabsTracer` is a subclass and `Tracer` a pure interface, or `KrabsTracer` is a "normal" class that `Tracer` forwards to.

pierricgimmig replied:
I want to make sure that even if a user holds on to a Tracer object for multiple Start/Stop cycles, a new KrabsTracer object will be created for each session, avoiding the need for resetting the KrabsTracer explicitly, and clearing the memory on session Stop. But yeah, I agree, the current setup is weird, will rework.

pierricgimmig replied:
Introduced a TracerInterface in WindowsTracing. Both `Tracer` and `KrabsTracer` implement the interface. Tracer also owns an opaque pointer to a `TracerInterface`, which internally is set to a KrabsTracer instance on "Start". This is done to hide the implementation and also make sure that a new KrabsTracer object is created on every profiling session, and that we clear its memory on "Stop".

florian-kuebler commented on src/WindowsTracing/KrabsTracer.h:
```diff
@@ -0,0 +1,45 @@
+
+namespace orbit_windows_tracing {
+
+class KrabsTracer : public Tracer {
```
Please add a brief class level comment.

pierricgimmig replied:
Done, thanks.

florian-kuebler commented on src/WindowsTracing/KrabsTracer.cpp:
```diff
@@ -0,0 +1,109 @@
+
+using orbit_grpc_protos::SchedulingSlice;
+
+KrabsTracer::KrabsTracer(orbit_grpc_protos::CaptureOptions capture_options,
```
I'd love to see a test for this class.

pierricgimmig replied:
I looked into it a little, it's not easily unit-testable. We'd have to essentially generate fake ETW data, which is a bit of work. We can definitely go for an integration test however, I have one locally. I'll dig deeper for the unit test, but this will be in another PR.

florian-kuebler replied:
Okay, thanks for looking into this.

florian-kuebler commented:
Very cool PR. Looking forward to see it in.

dpallotti commented on src/WindowsTracing/KrabsTracer.cpp:
```diff
@@ -0,0 +1,109 @@
+      context_switch_manager_->ProcessCpuEvent(cpu, old_tid, new_tid, timestamp_ns);
+    } break;
+    default:
+      break;
```
Can this happen or should it be `UNREACHABLE()`?

dpallotti commented on src/WindowsTracing/KrabsTracer.cpp:
```diff
@@ -0,0 +1,109 @@
+  LOG("Free buffers: %u", trace_stats.buffersFree);
+  LOG("Buffers written: %u", trace_stats.buffersWritten);
+  LOG("Buffers lost: %u", trace_stats.buffersLost);
+  LOG("Events total (handled+lost): %lu", trace_stats.eventsTotal);
+  LOG("Events handled: %lu", trace_stats.eventsHandled);
```
```suggestion
  LOG("Events total (handled+lost): %u", trace_stats.eventsTotal);
  LOG("Events handled: %u", trace_stats.eventsHandled);
```
Just consistency.

pierricgimmig replied:
Done, thanks.

dpallotti commented on src/WindowsTracing/KrabsTracer.cpp:
```diff
@@ -0,0 +1,109 @@
+    : Tracer(std::move(capture_options), listener), trace_(KERNEL_LOGGER_NAME) {
+  SetTraceProperties();
+  EnableProviders();
+};
```
```suggestion
}
```

pierricgimmig replied:
Done, thanks.

dpallotti commented on src/WindowsTracing/ContextSwitchManager.h:
```diff
@@ -0,0 +1,58 @@
+// events. It assumes that context switches for the same core come in order.
+class ContextSwitchManager {
+ public:
+  explicit ContextSwitchManager(TracerListener* listener) : listener_(listener){};
```
```suggestion
  explicit ContextSwitchManager(TracerListener* listener) : listener_(listener) {}
```

pierricgimmig replied:
Done, thanks.

dpallotti commented on src/WindowsTracing/KrabsTracer.h:
```diff
@@ -0,0 +1,45 @@
+
+ private:
+  std::unique_ptr<ContextSwitchManager> context_switch_manager_;
+  std::unique_ptr<std::thread> thread_;
```
Maybe a more descriptive name? `kernel_trace_thread` or `trace_thread`?

pierricgimmig replied:
Yeah good idea, thanks, done.

dpallotti commented on src/WindowsTracing/EventTypes.h:
```diff
@@ -0,0 +1,91 @@
+
+// https://docs.microsoft.com/en-us/windows/win32/etw/thread-typegroup1
+// [EventType{ 1, 2, 3, 4 }, EventTypeName{ "Start", "End", "DCStart", "DCEnd" }]
+struct Thread_TypeGroup1 {
```
I'm a bit lost with this file, some explanatory comment might be necessary.
For example I'm confused by the structs. In the links they have fields but here you don't need them thanks to `krabs`? But then you use `struct`s of the same name as containers of the opcodes. I can see the point of using the same name but it's also confusing because then I expect the same exact `struct`, maybe a suffix would be useful? Or do you even need the structs if the opcodes are all distinct?

pierricgimmig replied:
Replaced `struct` by `namespace`. I like grouping them by event type better than having a flat set of opcodes. Also, added a comment.

pierricgimmig commented on src/WindowsTracing/ClockUtils.h:
```diff
@@ -0,0 +1,39 @@
+  ClockUtils() {
+    LARGE_INTEGER frequency;
+    QueryPerformanceFrequency(&frequency);
+    performance_frequency_ = frequency.QuadPart;
```
Simplified it, thanks.

pierricgimmig commented on src/WindowsTracing/ClockUtils.h:
```diff
@@ -0,0 +1,39 @@
+
+class ClockUtils {
+ public:
+  [[nodiscard]] static inline uint64_t RawTimestampToNs(uint64_t raw_timestamp) {
```
removed the class altogether.

pierricgimmig commented on src/WindowsTracing/ClockUtils.h:
```diff
@@ -0,0 +1,39 @@
+
+namespace orbit_windows_tracing {
+
+class ClockUtils {
```
Replaced class by functions whose names are self-explanatory.

pierricgimmig commented on src/WindowsTracing/ContextSwitchManager.cpp:
```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
```
Done, thanks.

pierricgimmig commented on src/WindowsTracing/ContextSwitchManager.h:
```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
```
Done, thanks.

dpallotti replied:
I guess you missed this one :D

pierricgimmig replied:
fixed, thanks.

pierricgimmig commented on src/WindowsTracing/ContextSwitchManager.h:
```diff
@@ -0,0 +1,58 @@
+
+  [[nodiscard]] const Stats& GetStats() { return stats_; }
+
+ protected:
```
done, thanks.

pierricgimmig commented on src/WindowsTracing/include/WindowsTracing/TracerListener.h:
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
```
Done, thanks.

pierricgimmig commented on src/WindowsTracing/ContextSwitchManager.h:
```diff
@@ -0,0 +1,58 @@
+
+namespace orbit_windows_tracing {
+
+// For each core, keeps the last context switch into a process and matches it
+// with the next context switch away from a process to produce SchedulingSlice
```
Done, thanks.

pierricgimmig commented on src/WindowsTracing/ContextSwitchManager.h:
```diff
@@ -0,0 +1,58 @@
+
+// For each core, keeps the last context switch into a process and matches it
+// with the next context switch away from a process to produce SchedulingSlice
+// events. It assumes that context switches for the same core come in order.
```
Yep, added a mention in the comment.

pierricgimmig commented on src/WindowsTracing/ContextSwitchManager.h:
```diff
@@ -0,0 +1,58 @@
+  ContextSwitchManager() = delete;
+
+  void ProcessThreadEvent(uint32_t tid, uint32_t pid);
+  void ProcessCpuEvent(uint16_t cpu, uint32_t old_tid, uint32_t new_tid, uint64_t timestamp_ns);
```
Renamed both methods.

pierricgimmig commented:
@florian-kuebler , @dpallotti , thanks for the review! I addressed your comments, but will add more tests tomorrow, probably in another PR. PTAL. Thanks!

dpallotti commented on src/WindowsTracing/ClockUtils.h:
```diff
@@ -0,0 +1,27 @@
+
+#include <cstdint>
+
+namespace orbit_windows_tracing::clock_utils {
```
We agree not to use nested namespaces: https://github.com/google/orbit/blob/main/DEVELOPMENT.md#namespaces
In this case you can probably just remove the inner one.

pierricgimmig replied:
Done.

dpallotti commented on src/WindowsTracing/include/WindowsTracing/Tracer.h:
```diff
@@ -0,0 +1,31 @@
+namespace orbit_windows_tracing {
+
+// Forwards interesting Windows kernel events to a TraceListener.
+class Tracer : public TracerInterface {
```
So again on this class, this is still strange to me.
My understanding is that you have `TracerInterface` so that `KrabsTracer` doesn't have to be exposed in the `include` directory and can be instantiated in the `.cpp` file. So far so good.
But you also have `Tracer` implement this interface and that is what I'm confused about: You have a `TracerInterface` (`Tracer`) contain another `TracerInterface` (`KrabsTracer`), which is a bit strange; And `TracerInterface` now has two usages, one exposing `Tracer` as `TracerInterface`, one hiding `KrabsTracer` behind `TracerInterface`.
As I don't currently see a reason for `Tracer` to implement `TracerInterface`, only `KrabsTracer` should implement it.

So in the end I think that `public TracerInterface` and the two `override`s would simply be removed.
As a result, `TracerInterface` would be internal to the module and should end up in `orbit_windows_tracing_internal` as it's in the `include` directory, again as per https://github.com/google/orbit/blob/main/DEVELOPMENT.md#namespaces

florian-kuebler replied:
+1. Basically simply drop the inheritance of TracerInterface here and that's it.

pierricgimmig replied:
Both `Tracer` and `KrabsTracer` need to store the CaptureOptions and TracerListener. Having them both implement the `TracerInterface` puts that logic at one place. Also, both can be used interchangeably through the `TracerInterface`. 

Removing the inheritance in Tracer would mean duplicating code for storing `CaptureOptions` and `TracerListener`, and losing the property that `Tracer` could be thought of as just an implementation of the `TracerInterface`, which I find nice as a user as the interface acts as sort of lean documentation. Also, we could imagine exposing different implementations of the TracerInterface publicly, in which case we need a public interface.

pierricgimmig replied:
@florian-kuebler , ok I ended up removing the TracerInterface in favor of a TracerImpl.h/.cpp the hierarchy, ptal.

dpallotti commented on src/WindowsTracing/Tracer.cpp:
```diff
@@ -0,0 +1,28 @@
+}
+
+void Tracer::Stop() {
+  if (tracer_impl_ != nullptr) {
```
Should this be a `CHECK`?

pierricgimmig replied:
Done.

dpallotti commented on src/WindowsTracing/EtwEventTypes.h:
```diff
@@ -0,0 +1,88 @@
+
+#include <cstdint>
+
+namespace orbit_windows_tracing::etw {
+
+// ETW event types grouped by event class. Note that an event type name is of the form
+// "[parent class]_[event class]". The parent class is associated with a guid that is used
+// to dispatch events to a particular callback for processing.
+// See https://docs.microsoft.com/en-us/windows/win32/etw/event-tracing-mof-classes.
+
+// https://docs.microsoft.com/en-us/windows/win32/etw/thread-typegroup1
+namespace Thread_TypeGroup1 {
```
So again, we decided not to use nested namespaces and I'd like to avoid exceptions (even more as `orbit_windows_tracing::etw` could be mistaken with a possible `etw` and the inner ones can't follow snake_case).

`struct`s are fine for grouping, just I was confused by the homonymy. What about just a suffix? For example:
`struct Thread_TypeGroup1_Constants`, or `struct Thread_TypeGroup1_Ids`, or `struct Thread_TypeGroup1_EventTypeIds`, ...?

pierricgimmig replied:
Went for flat list of constants.

dpallotti commented on src/WindowsTracing/EtwEventTypes.h:
```diff
@@ -0,0 +1,88 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef WINDOWS_TRACING_EVENT_TYPES_H_
+#define WINDOWS_TRACING_EVENT_TYPES_H_
```
Update the include guards.

pierricgimmig replied:
Done, thanks.

florian-kuebler commented on src/WindowsTracing/MockTracerListener.h:
```diff
@@ -0,0 +1,20 @@
+#include "WindowsTracing/TracerListener.h"
+#include "capture.pb.h"
+namespace orbit_windows_tracing {
+class MockTracerListener : public orbit_windows_tracing::TracerListener {
```
Personal preference: You can move that mock class into the test directly.

pierricgimmig replied:
Yep, thanks, done.

florian-kuebler commented on src/WindowsTracing/include/WindowsTracing/TracerListener.h:
```diff
@@ -0,0 +1,20 @@
+
+namespace orbit_windows_tracing {
+
+class TracerListener {
```
Could you also add a class level comment here?

pierricgimmig replied:
Done.

florian-kuebler commented on src/WindowsTracing/include/WindowsTracing/TracerInterface.h:
```diff
@@ -0,0 +1,27 @@
+
+namespace orbit_windows_tracing {
+
+class TracerInterface {
```
Can you add a class level comment here?

pierricgimmig replied:
Done.

florian-kuebler commented on src/WindowsTracing/Tracer.cpp:
```diff
@@ -0,0 +1,28 @@
+
+void Tracer::Start() {
+  CHECK(tracer_impl_ == nullptr);
+  tracer_impl_ = std::make_unique<KrabsTracer>(capture_options_, listener_);
```
If at any point there would be another TracerInterface implementation, we probably want some kind of factory and a generic way to instantiate the concrete tracer here (e.g. as template). But for now, we can keep it as it is.

pierricgimmig commented on src/WindowsTracing/ContextSwitchManagerTest.cpp:
```diff
@@ -0,0 +1,109 @@
+  manager.ProcessCpuEvent(/*cpu=*/0, /*old_tid=*/2, /*new_tid=*/1, /*timestamp_ns=*/1);
+
+  EXPECT_EQ(fake_listener.scheduling_slices_.size(), 1);
+  EXPECT_EQ(fake_listener.scheduling_slices_[0].pid(), 3);
```
Done in TEST(ContextSwitch, ValidSchedulingSlice)

florian-kuebler reviewed (approved):
LGTM, thanks!
